pyro/contrib/zuko.py

```diff
@@ -0,0 +1,46 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+import torch
+from torch import Size, Tensor
+
+import pyro
+
+
+class Zuko2Pyro(pyro.distributions.TorchDistribution):
+    r"""Wraps a Zuko (or PyTorch) distribution as a Pyro distribution."""
+
+    def __init__(self, dist: torch.distributions.Distribution):
+        self.dist = dist
+        self.cache = {}
+
+    @property
+    def has_rsample(self) -> bool:
+        return self.dist.has_rsample
+
+    @property
+    def event_shape(self) -> Size:
+        return self.dist.event_shape
+
+    @property
+    def batch_shape(self) -> Size:
+        return self.dist.batch_shape
+
+    def __call__(self, shape: Size = ()) -> Tensor:
+        if hasattr(self.dist, "rsample_and_log_prob"):  # fast sampling + scoring
+            x, self.cache[x] = self.dist.rsample_and_log_prob(shape)
+        elif self.has_rsample:
+            x = self.dist.rsample(shape)
+        else:
+            x = self.dist.sample(shape)
+
+        return x
+
+    def log_prob(self, x: Tensor) -> Tensor:
+        if x in self.cache:
+            return self.cache[x]
```

Comment:
@fritzo can you please take a look at the caching logic?

Reply:
Note that I don't clear the cache as I consider that `Zuko2Pyro` objects should be temporary.

Reply:
Note that the `cache` is only used when `x` is sampled with `rsample_and_log_prob`. But, yes I will mention it in the docstring.